### ISSUE TITLE
add_dynamic_page_feed Example not working

### ISSUE THREAD

**Issue by skiwan** [googleads/google-ads-python#312]
Hey, Im currently trying to add a custom page feed to our dynamic search adds campaigns and let google ads only use the supplied urls.

Im currently using the google-ads-python api version 3

I have been trying the provided example add_dynamic_page_feed, unfortunately it is not working.

### The problem

When using the current provided example code from here "https://developers.google.com/google-ads/api/docs/dynamic-search-ads/create-dynamic-search-ads#python_1"  one runs into the already known issue that with the update function the service wants the domain name value. I have added the domain name value to the example code (update function) I receive an operation that looks ok, the mutate operation goes trough without any errors but the settings for the campaign dont change.

Next I tried the example that is currently in github "https://github.com/googleads/google-ads-python/blob/master/examples/advanced_operations/add_dynamic_page_feed.py"

When I use 
updated_campaign.dynamic_search_ads_setting.feeds.append(feed_details.resource_name)

I receive the following error
"updated_campaign.dynamic_search_ads_setting.feeds.append(feed_resource_name)
  File "E:\MerlinsProgrammingPlayground\env\lib\site-packages\google\protobuf\internal\containers.py", line 388, in append
    new_element.CopyFrom(value)
  File "E:\MerlinsProgrammingPlayground\env\lib\site-packages\google\protobuf\message.py", line 118, in CopyFrom
    self.MergeFrom(other_msg)
  File "E:\MerlinsProgrammingPlayground\env\lib\site-packages\google\protobuf\internal\python_message.py", line 1302, in MergeFrom
    raise TypeError(
TypeError: Parameter to MergeFrom() must be instance of same class: expected StringValue got str."

The only thing Im doing different to the provided example is that I am not programatically create a new feed. Instead I am querying an already existing one the following way

"def get_page_feed(feed_name):
    query = """
    SELECT
    feed.id,
    feed.resource_name,
    feed.attributes
        FROM
            feed
        where
            feed.name = "{}"
        """.format(feed_name)
        response = ga_service.search(google_ads_id, query=query)
        # Maps specific fields in each row in the response to a dict. This would
        # overwrite the same fields in the dict for each row, but we know we'll
        # only one row will be returned.
        for row in response:
            resource_name = row.feed.resource_name
            feed_id = row.feed.id.value
            print(row.feed, 'feed')
        return {
            'resource_name': resource_name,
            'id' : feed_id
        }"
With this I get our current feed and the resource name as a string which looks like  customers/12345667/campaigns/12345667 (random numbers used here as example)

If Instead of the append function I use feed = ...add() and then feed.value = resource_name the mutate operation also goes through the service without any errors but still no update within the campaign settings.

How can we solve this?
If any additional info is required please let me know

**Comment by BenRKarl** [comment]
Hi @skiwan I believe the issue here is that the example uses `v5` of the API, where a number of protos, including `DynamicSearchAdsSetting` have had their fields changed from wrapper types to primitive types. In `v5` [`DynamicSearchAdsSetting.feeds`](https://github.com/googleapis/googleapis/blob/master/google/ads/googleads/v5/resources/campaign.proto#L100) is a `repeated string` field, so in the example we append a string directly, for example:

```py
campaign.dynamic_search_ads_setting.feeds.append('foo bar')
```

In `v3` of the API [`DynamicSearchAdsSetting.feeds`](https://github.com/googleapis/googleapis/blob/master/google/ads/googleads/v3/resources/campaign.proto#L132) is a `repeated StringValue` field, so a `google.protobuf.wrappers_pb2.StringValue` instance must be passed in, not a native Python `str`. So for your use case you would need to do either of the following:

```py
# Initialize a StringValue instance using the client
string_value = client.get_type("StringValue", version="v3")
string_value.value = "foo bar"
campaign.dymanic_search_ads_setting.feeds.append(string_value)

# Initialize a StringValue via the repeated field's add method
string_value = campaign.dynamic_search_ads_setting.feeds.add()
string_value.value = "foo bar"
```

In the past our examples have used the second pattern.

Now you mentioned you're using `StringValue`and still not seeing the field get updated, is that right? If that's the case then it might be the result of a bug. Do you receive a successful response in those requests? If so can you enable logging and share a request_id?

**Comment by skiwan** [comment]
Thanks for the response @BenRKarl ,

Ive done some elaborate testing today.
First the feeds.add() (second pattern) I had tried earlier and the operation goes trough without any errors but the settings on the campaing are not changed afterwards.

In regards to using the string value Im getting the error that a feed id is missing.

I have tried nearly all examples and ways to add a feed each while still creating the campaign (create operation) and after creating the campaign within an update (update operation)


Following examples of what I tried and the corresponding error codes.

## Feed update Test (campaign status is always PAUSED for all examples given)


### within Campaign Creation use 

```python
	string_value = client.get_type('StringValue', version='v3')
	string_value.value = feed_resource_name

	campaign.dynamic_search_ads_setting.domain_name.value = domain_name
	campaign.dynamic_search_ads_setting.language_code.value = sales_country
	campaign.dynamic_search_ads_setting.feeds.append(string_value)
	campaign.dynamic_search_ads_setting.use_supplied_urls_only.value = True
```

#### Leads to 

```
Request made: ClientCustomerId: ********* , Host: googleads.googleapis.com:443, Method: /google.ads.googleads.v3.services.CampaignService/MutateCampaigns, RequestId: lbnbPQcXUUqUIAdNzN3bvw, IsFault: True, FaultMessage: At least one feed id should be present.
Request with ID "lbnbPQcXUUqUIAdNzN3bvw" failed with status "INVALID_ARGUMENT" and includes the following errors:
        Error with message "At least one feed id should be present.".
                On field: operations
                On field: create
                On field: dynamic_search_ads_setting
Traceback (most recent call last):
  File "dsa_campaign_script.py", line 735, in <module>
    campaign_resource_name = create_dsa_campaign(client, google_ads_id,google_merchant_center_account_id, campaign_name, budget_ressource, True, False, campaign_service, sales_country, domain, feed_ressource_name,feed_id)
  File "dsa_campaign_script.py", line 304, in create_dsa_campaign
    campaign_resource_name = campaign_response.results[0].resource_name
UnboundLocalError: local variable 'campaign_response' referenced before assignment
```
_______________________________


### Within Update after Campaign was created

```python
	campaign_operation = client.get_type('CampaignOperation', version='v3')
	campaign_operation.update.CopyFrom(campaign)
	updated_campaign = campaign_operation.update
	
	string_value = client.get_type('StringValue', version='v3')
	string_value.value = feed_resource_name
	updated_campaign.dynamic_search_ads_setting.feeds.append(string_value)

	field_mask = google.api_core.protobuf_helpers.field_mask(campaign, updated_campaign)
	updated_campaign.dynamic_search_ads_setting.use_supplied_urls_only.value = True
	updated_campaign.dynamic_search_ads_setting.domain_name.value = domain_name
	updated_campaign.dynamic_search_ads_setting.language_code.value = sales_country
	campaign_operation.update_mask.CopyFrom(field_mask)
```
#### Leads to

	Goes through, no error message
		response = campaign_service.mutate_campaigns(google_ads_id,
												 [campaign_operation])
		resource_name = response.results[0].resource_name

	resource_name returned here #customers/9465697505/campaigns/11126425128 == Resource name of the campaign that was supposed to update

	But settings of campaign are not updated, when opened in Google ads no changes to campaign settings

__________________________________

All of these examples happen BEFORE I add Adgroups to the campaign, could this be a problem?


__________________________________

### Within Update after Campaign was created AND filled

```python
	campaign_operation = client.get_type('CampaignOperation', version='v3')
	# Copy the retrieved campaign onto the new campaign operation.
	campaign_operation.update.CopyFrom(campaign)
	updated_campaign = campaign_operation.update
	
	string_value = client.get_type('StringValue', version='v3')
	string_value.value = feed_resource_name
	updated_campaign.dynamic_search_ads_setting.feeds.append(string_value)

	field_mask = google.api_core.protobuf_helpers.field_mask(campaign, updated_campaign)
	updated_campaign.dynamic_search_ads_setting.use_supplied_urls_only.value = True
	updated_campaign.dynamic_search_ads_setting.domain_name.value = domain_name
	updated_campaign.dynamic_search_ads_setting.language_code.value = sales_country
	campaign_operation.update_mask.CopyFrom(field_mask)

	print(campaign_operation, 'operation')
	print(feed_resource_name, 'feed_ressource')
	# Submit the campaign operation and update the campaign.
	response = campaign_service.mutate_campaigns(google_ads_id,
												 [campaign_operation])
	resource_name = response.results[0].resource_name
```

#### Leads to

```
Request made: ClientCustomerId: 9465697505, Host: googleads.googleapis.com:443, Method: /google.ads.googleads.v3.services.CampaignService/MutateCampaigns, RequestId: yjQ5AV6BlQYrXWKV9PlXtw, IsFault: True, FaultMessage: The required field was not present.
Traceback (most recent call last):
  File "dsa_campaign_script.py", line 746, in <module>
    give_feed_settings_to_campaign(client, ga_service, google_ads_id,campaign_service,campaign_resource_name,domain,sales_country, feed_ressource_name)
  File "dsa_campaign_script.py", line 249, in give_feed_settings_to_campaign
    response = campaign_service.mutate_campaigns(google_ads_id,
  File "E:\MerlinsProgrammingPlayground\env\lib\site-packages\google\ads\google_ads\v3\services\campaign_service_client.py", line 297, in mutate_campaigns
    return self._inner_api_calls['mutate_campaigns'](request, retry=retry, timeout=timeout, metadata=metadata)
  File "E:\MerlinsProgrammingPlayground\env\lib\site-packages\google\api_core\gapic_v1\method.py", line 143, in __call__
    return wrapped_func(*args, **kwargs)
  File "E:\MerlinsProgrammingPlayground\env\lib\site-packages\google\api_core\retry.py", line 281, in retry_wrapped_func
    return retry_target(
  File "E:\MerlinsProgrammingPlayground\env\lib\site-packages\google\api_core\retry.py", line 184, in retry_target
    return target()
  File "E:\MerlinsProgrammingPlayground\env\lib\site-packages\google\api_core\timeout.py", line 214, in func_with_timeout
    return func(*args, **kwargs)
  File "E:\MerlinsProgrammingPlayground\env\lib\site-packages\google\api_core\grpc_helpers.py", line 57, in error_remapped_callable
    return callable_(*args, **kwargs)
  File "E:\MerlinsProgrammingPlayground\env\lib\site-packages\grpc\_interceptor.py", line 216, in __call__
    response, ignored_call = self._with_call(request,
  File "E:\MerlinsProgrammingPlayground\env\lib\site-packages\grpc\_interceptor.py", line 257, in _with_call
    return call.result(), call
  File "E:\MerlinsProgrammingPlayground\env\lib\site-packages\grpc\_interceptor.py", line 126, in result
    raise self._exception
  File "E:\MerlinsProgrammingPlayground\env\lib\site-packages\grpc\_interceptor.py", line 241, in continuation
    response, call = self._thunk(new_method).with_call(
  File "E:\MerlinsProgrammingPlayground\env\lib\site-packages\grpc\_interceptor.py", line 266, in with_call
    return self._with_call(request,
  File "E:\MerlinsProgrammingPlayground\env\lib\site-packages\grpc\_interceptor.py", line 257, in _with_call
    return call.result(), call
  File "E:\MerlinsProgrammingPlayground\env\lib\site-packages\grpc\_interceptor.py", line 126, in result
    raise self._exception
  File "E:\MerlinsProgrammingPlayground\env\lib\site-packages\grpc\_interceptor.py", line 241, in continuation
    response, call = self._thunk(new_method).with_call(
  File "E:\MerlinsProgrammingPlayground\env\lib\site-packages\grpc\_interceptor.py", line 266, in with_call
    return self._with_call(request,
  File "E:\MerlinsProgrammingPlayground\env\lib\site-packages\grpc\_interceptor.py", line 254, in _with_call
    call = self._interceptor.intercept_unary_unary(continuation,
  File "E:\MerlinsProgrammingPlayground\env\lib\site-packages\google\ads\google_ads\interceptors\exception_interceptor.py", line 168, in intercept_unary_unary
    self._handle_grpc_failure(response)
  File "E:\MerlinsProgrammingPlayground\env\lib\site-packages\google\ads\google_ads\interceptors\exception_interceptor.py", line 140, in _handle_grpc_failure
    raise self._get_error_from_response(response)
google.ads.google_ads.errors.GoogleAdsException: (<_InactiveRpcError of RPC that terminated with:
        status = StatusCode.INVALID_ARGUMENT
        details = "Request contains an invalid argument."
        debug_error_string = "{"created":"@1600168140.273000000","description":"Error received from peer ipv4:172.217.21.170:443","file":"src/core/lib/surface/call.cc","file_line":1056,"grpc_message":"Request contains an invalid argument.","grpc_status":3}"
>, <_InactiveRpcError of RPC that terminated with:
        status = StatusCode.INVALID_ARGUMENT
        details = "Request contains an invalid argument."
        debug_error_string = "{"created":"@1600168140.273000000","description":"Error received from peer ipv4:172.217.21.170:443","file":"src/core/lib/surface/call.cc","file_line":1056,"grpc_message":"Request contains an invalid argument.","grpc_status":3}"
>, errors {
  error_code {
    field_error: REQUIRED
  }
  message: "The required field was not present."
  trigger {
    string_value: ""
  }
  location {
    field_path_elements {
      field_name: "operations"
      index {
      }
    }
    field_path_elements {
      field_name: "update"
    }
    field_path_elements {
      field_name: "dynamic_search_ads_setting"
    }
    field_path_elements {
      field_name: "domain_name"
    }
  }
}
errors {
  error_code {
    field_error: REQUIRED
  }
  message: "The required field was not present."
  trigger {
    string_value: ""
  }
  location {
    field_path_elements {
      field_name: "operations"
      index {
      }
    }
    field_path_elements {
      field_name: "update"
    }
    field_path_elements {
      field_name: "dynamic_search_ads_setting"
    }
    field_path_elements {
      field_name: "language_code"
    }
  }
}
, 'yjQ5AV6BlQYrXWKV9PlXtw')Request made: ClientCustomerId: 9465697505, Host: googleads.googleapis.com:443, Method: /google.ads.googleads.v3.services.CampaignService/MutateCampaigns, RequestId: yjQ5AV6BlQYrXWKV9PlXtw, IsFault: True, FaultMessage: The required field was not present.
Traceback (most recent call last):
  File "dsa_campaign_script.py", line 746, in <module>
    give_feed_settings_to_campaign(client, ga_service, google_ads_id,campaign_service,campaign_resource_name,domain,sales_country, feed_ressource_name)
  File "dsa_campaign_script.py", line 249, in give_feed_settings_to_campaign
    response = campaign_service.mutate_campaigns(google_ads_id,
  File "E:\MerlinsProgrammingPlayground\env\lib\site-packages\google\ads\google_ads\v3\services\campaign_service_client.py", line 297, in mutate_campaigns
    return self._inner_api_calls['mutate_campaigns'](request, retry=retry, timeout=timeout, metadata=metadata)
  File "E:\MerlinsProgrammingPlayground\env\lib\site-packages\google\api_core\gapic_v1\method.py", line 143, in __call__
    return wrapped_func(*args, **kwargs)
  File "E:\MerlinsProgrammingPlayground\env\lib\site-packages\google\api_core\retry.py", line 281, in retry_wrapped_func
    return retry_target(
  File "E:\MerlinsProgrammingPlayground\env\lib\site-packages\google\api_core\retry.py", line 184, in retry_target
    return target()
  File "E:\MerlinsProgrammingPlayground\env\lib\site-packages\google\api_core\timeout.py", line 214, in func_with_timeout
    return func(*args, **kwargs)
  File "E:\MerlinsProgrammingPlayground\env\lib\site-packages\google\api_core\grpc_helpers.py", line 57, in error_remapped_callable
    return callable_(*args, **kwargs)
  File "E:\MerlinsProgrammingPlayground\env\lib\site-packages\grpc\_interceptor.py", line 216, in __call__
    response, ignored_call = self._with_call(request,
  File "E:\MerlinsProgrammingPlayground\env\lib\site-packages\grpc\_interceptor.py", line 257, in _with_call
    return call.result(), call
  File "E:\MerlinsProgrammingPlayground\env\lib\site-packages\grpc\_interceptor.py", line 126, in result
    raise self._exception
  File "E:\MerlinsProgrammingPlayground\env\lib\site-packages\grpc\_interceptor.py", line 241, in continuation
    response, call = self._thunk(new_method).with_call(
  File "E:\MerlinsProgrammingPlayground\env\lib\site-packages\grpc\_interceptor.py", line 266, in with_call
    return self._with_call(request,
  File "E:\MerlinsProgrammingPlayground\env\lib\site-packages\grpc\_interceptor.py", line 257, in _with_call
    return call.result(), call
  File "E:\MerlinsProgrammingPlayground\env\lib\site-packages\grpc\_interceptor.py", line 126, in result
    raise self._exception
  File "E:\MerlinsProgrammingPlayground\env\lib\site-packages\grpc\_interceptor.py", line 241, in continuation
    response, call = self._thunk(new_method).with_call(
  File "E:\MerlinsProgrammingPlayground\env\lib\site-packages\grpc\_interceptor.py", line 266, in with_call
    return self._with_call(request,
  File "E:\MerlinsProgrammingPlayground\env\lib\site-packages\grpc\_interceptor.py", line 254, in _with_call
    call = self._interceptor.intercept_unary_unary(continuation,
  File "E:\MerlinsProgrammingPlayground\env\lib\site-packages\google\ads\google_ads\interceptors\exception_interceptor.py", line 168, in intercept_unary_unary
    self._handle_grpc_failure(response)
  File "E:\MerlinsProgrammingPlayground\env\lib\site-packages\google\ads\google_ads\interceptors\exception_interceptor.py", line 140, in _handle_grpc_failure
    raise self._get_error_from_response(response)
google.ads.google_ads.errors.GoogleAdsException: (<_InactiveRpcError of RPC that terminated with:
        status = StatusCode.INVALID_ARGUMENT
        details = "Request contains an invalid argument."
        debug_error_string = "{"created":"@1600168140.273000000","description":"Error received from peer ipv4:172.217.21.170:443","file":"src/core/lib/surface/call.cc","file_line":1056,"grpc_message":"Request contains an invalid argument.","grpc_status":3}"
>, <_InactiveRpcError of RPC that terminated with:
        status = StatusCode.INVALID_ARGUMENT
        details = "Request contains an invalid argument."
        debug_error_string = "{"created":"@1600168140.273000000","description":"Error received from peer ipv4:172.217.21.170:443","file":"src/core/lib/surface/call.cc","file_line":1056,"grpc_message":"Request contains an invalid argument.","grpc_status":3}"
>, errors {
  error_code {
    field_error: REQUIRED
  }
  message: "The required field was not present."
  trigger {
    string_value: ""
  }
  location {
    field_path_elements {
      field_name: "operations"
      index {
      }
    }
    field_path_elements {
      field_name: "update"
    }
    field_path_elements {
      field_name: "dynamic_search_ads_setting"
    }
    field_path_elements {
      field_name: "domain_name"
    }
  }
}
errors {
  error_code {
    field_error: REQUIRED
  }
  message: "The required field was not present."
  trigger {
    string_value: ""
  }
  location {
    field_path_elements {
      field_name: "operations"
      index {
      }
    }
    field_path_elements {
      field_name: "update"
    }
    field_path_elements {
      field_name: "dynamic_search_ads_setting"
    }
    field_path_elements {
      field_name: "language_code"
    }
  }
}
, 'yjQ5AV6BlQYrXWKV9PlXtw')```
__________________________________


Maybe any of this helps you or you can already see whats going wrong.

If not let me know what else you would need.

**Comment by skiwan** [comment]
@BenRKarl 

Hey Carl, now im the one with a late respone ;)
Tested today with v5 everything working fine. Will upgrade our current code to v5 which should solve our problem here. Thanks for helping out

Is there already a v6 or is v5 the newest?
After that info you can close the issue, thanks very much !

**Comment by BenRKarl** [comment]
@skiwan `v6` is the newest, and note that we just released version `10.0.0` of this client library, which includes a number of breaking interface improvements. There's more background in #400
